src/bootstrap/lib.rs

```diff
@@ -965,55 +963,97 @@ impl Build {
 
     /// Runs a command, printing out nice contextual information if it fails.
     fn run(&self, cmd: &mut Command) {
-        if self.config.dry_run() {
-            return;
-        }
-        self.verbose(&format!("running: {cmd:?}"));
-        run(cmd, self.is_verbose())
+        self.run_cmd(cmd);
     }
 
     /// Runs a command, printing out nice contextual information if it fails.
     fn run_quiet(&self, cmd: &mut Command) {
-        if self.config.dry_run() {
-            return;
-        }
-        self.verbose(&format!("running: {cmd:?}"));
-        run_suppressed(cmd)
+        let cmd: BootstrapCommand<'_> = cmd.into();
+        self.run_cmd(
+            cmd.failure_output(OutputOnFailure::Verbose).failure_behavior(BehaviorOnFailure::Exit),
+        );
     }
 
     /// Runs a command, printing out nice contextual information if it fails.
     /// Exits if the command failed to execute at all, otherwise returns its
     /// `status.success()`.
     fn run_quiet_delaying_failure(&self, cmd: &mut Command) -> bool {
-        if self.config.dry_run() {
-            return true;
-        }
-        if !self.fail_fast {
-            self.verbose(&format!("running: {cmd:?}"));
-            if !try_run_suppressed(cmd) {
-                let mut failures = self.delayed_failures.borrow_mut();
-                failures.push(format!("{cmd:?}"));
-                return false;
-            }
-        } else {
-            self.run_quiet(cmd);
-        }
-        true
+        let cmd: BootstrapCommand<'_> = cmd.into();
+        self.run_cmd(
+            cmd.failure_output(OutputOnFailure::Verbose)
+                .failure_behavior(BehaviorOnFailure::DelayFail),
+        )
     }
 
     /// Runs a command, printing out contextual info if it fails, and delaying errors until the build finishes.
     pub(crate) fn run_delaying_failure(&self, cmd: &mut Command) -> bool {
-        if !self.fail_fast {
-            #[allow(deprecated)] // can't use Build::try_run, that's us
-            if self.config.try_run(cmd).is_err() {
-                let mut failures = self.delayed_failures.borrow_mut();
-                failures.push(format!("{cmd:?}"));
-                return false;
+        let cmd: BootstrapCommand<'_> = cmd.into();
+        self.run_cmd(cmd.failure_behavior(BehaviorOnFailure::DelayFail))
+    }
+
+    /// A centralized function for running commands that do not return output.
+    pub(crate) fn run_cmd<'a, C: Into<BootstrapCommand<'a>>>(&self, cmd: C) -> bool {
```

Comment:
Shouldn't we handle this in `exec` module?

Reply:
That's not so easy yet - we need to access `self.fail_fast`, `self.delayed_failures`, `self.verbose` etc. It would need more refactoring to introduce some shared command context or something like that.